virtual_rainforest/models/abiotic_simple/simple_regression.py

```diff
@@ -0,0 +1,677 @@
+        },
+    )
+    output.append(atmospheric_co2)
+
+    # Precipitation at the surface is reduced as a function of leaf area index
+    precipitation_surface = data["precipitation"].isel(time=time_index) * (
+        1 - water_interception_factor * data["leaf_area_index"].sum(dim="layers")
+    )
+
+    # Mean soil moisture profile, [] and Runoff, [mm]
+    soil_moisture, surface_run_off = calculate_soil_moisture(
+        layer_roles=layer_roles,
+        precipitation_surface=precipitation_surface,
+        current_soil_moisture=data["soil_moisture"],
+        soil_moisture_capacity=soil_moisture_capacity,
+    )
+
+    output.append(soil_moisture)
+    output.append(surface_run_off)
+    return output
+
+
+# supporting functions
+def logarithmic(x: DataArray, gradient: float, intercept: float) -> DataArray:
+    """Logarithmic function.
+
+    Args:
+        x: x values
+        gradient: gradient
+        intercept: intercept
+
+    Returns:
+        y values
+    """
+    return DataArray(gradient * np.log(x) + intercept)
+
+
+def log_interpolation(
+    data: Data,
+    reference_data: DataArray,
+    leaf_area_index_sum: DataArray,
+    layer_roles: List,
+    layer_heights: DataArray,
+    upper_bound: float,
+    lower_bound: float,
+    gradient: float,
+) -> DataArray:
+    """LAI regression and logarithmic interpolation of variables for vertical profile.
+
+    Args:
+        data: Data object
+        reference_data: input variable at reference height
+        leaf_area_index_sum: leaf area index summed over all layers, [m m-1]
+        layer_roles: list of layer roles (soil, surface, subcanopy, canopy, above)
+        layer_heights: vertical layer heights, [m]
+        lower_bound: minimum allowed value
+        upper_bound: maximum allowed value
+        gradient: gradient of regression from :cite:t:`hardwick_relationship_2015`
+
+    Returns:
+        vertical profile of provided variable
+    """
+
+    # Calculate microclimatic variable at 1.5 m as function of leaf area index
+    lai_regression = DataArray(
+        leaf_area_index_sum * gradient + reference_data, dims="cell_id"
+    )
+
+    # Calculate per cell slope and intercept for logarithmic within-canopy profile
+    slope = (reference_data - lai_regression) / (
+        np.log(layer_heights.isel(layers=0)) - np.log(1.5)
+    )
+    intercept = lai_regression - slope * np.log(1.5)
+
+    # Calculate the values within cells by layer
+    layer_values = np.log(layer_heights) * slope + intercept
+
+    # set upper and lower bounds
+    layer_values_low = np.where(layer_values < lower_bound, lower_bound, layer_values)
+    return DataArray(
+        np.where(layer_values_low > upper_bound, upper_bound, layer_values_low),
+        dims=["layers", "cell_id"],
+        coords={
+            "layers": np.arange(0, len(layer_roles)),
+            "layer_roles": ("layers", layer_roles[0 : len(layer_roles)]),
+            "cell_id": data.grid.cell_id,
+        },
+    )
```

Comment:
You can use [`np.clip`](https://numpy.org/doc/stable/reference/generated/numpy.clip.html) function for this bound enforcement:
```suggestion
    return DataArray(
        np.clip(layer_values, lower_bound, upper_bound),
        dims=["layers", "cell_id"],
        coords={
            "layers": np.arange(0, len(layer_roles)),
            "layer_roles": ("layers", layer_roles[0 : len(layer_roles)]),
            "cell_id": data.grid.cell_id,
        },
    )
```
Also, I've seen in several places (here included) that you do `"layer_roles": ("layers", layer_roles[0 : len(layer_roles)])`. Why not simply `"layer_roles": ("layers", layer_roles)`?

Reply:
the `"layer_roles": ("layers", layer_roles[0 : len(layer_roles)])` is also an artifact from the building phase, I will remove it throughout the scipt